donkeycar/parts/gps.py

```diff
@@ -0,0 +1,641 @@
+import argparse
+from functools import reduce
+import logging
+import operator
+import threading
+import time
+
+import pynmea2
+import serial
+import utm
+
+logger = logging.getLogger("gps")
+logger.setLevel(logging.DEBUG)
+
+class gps:
+    def __init__(self, serial:str, baudrate:int = 9600, timeout:float = 0.5, debug = False):
+        self.serial = serial
+        self.baudrate = baudrate
+        self.timeout = timeout
+        self.debug = debug
+        self.positions = []  # tuple of (timestamp, longitude, latitude)
+        self.gps = None
+        self.lock = threading.Lock()
+        self.running = True
+        self._open()
+        self.clear()
+
+    def _open(self):
+        with self.lock:
+            self.gps = serial.Serial(self.serial, baudrate=self.baudrate, timeout=self.timeout)
+
+    def clear(self):
+        """
+        Clear the positions buffer
+        """
+        if self.gps is not None:
+            with self.lock:
+                try:
+                    if self.gps is not None and self.gps.is_open:
```

Comment:
In this case, I checked self.gps as a way to avoid doing the lock because it is expensive, but there is a race condition, so I need to check again after I do the lock.

Reply:
Thanks for catching this.